### DISCUSSION TITLE
root-level files in .ignore file not excluded

### DISCUSSION THREAD

**Discussion by victoriastuart** [BurntSushi/ripgrep#2250] [Q&A] [answered]
Expect: per the following example, file "apple" in root-level directory (relative to ripgrep command execution) should be ignored.

Issue: file 'apple' is not ignored (ditto re: "/composer.json", other examples - not shown here).

```bash
[victoria@victoria persagen]$ pwd; ls -l | grep apple
/mnt/Vancouver/domains/persagen
-rw-r--r--  1 victoria victoria      18 Jun 28 08:28 apple

[victoria@victoria persagen]$ cat apple
ohC8Ohfi
akaiv8Ee

[victoria@victoria persagen]$ cat .ignore
/apple
**/docs/*
**/files/*
**/versions/*

[victoria@victoria persagen]$ rg * --ignore-file .ignore -in -e 'ohC8Ohfi'
apple
1:ohC8Ohfi

[victoria@victoria persagen]$ uname -a    ## Arch Linux
Linux victoria 5.18.7-arch1-1 #1 SMP PREEMPT_DYNAMIC Sat, 25 Jun 2022 20:22:01 +0000 x86_64 GNU/Linux

[victoria@victoria persagen]$ which rg
/usr/bin/rg

[victoria@victoria persagen]$ rg --version
ripgrep 13.0.0
-SIMD -AVX (compiled)
+SIMD +AVX (runtime)

[victoria@victoria persagen]$ 

```

**Comment by BurntSushi** [comment, marked as answer]
Your search command, `rg * --ignore-file .ignore -in -e 'ohC8Ohfi'`, is telling ripgrep to explicitly search the `apple` file. So no, ripgrep isn't going to ignore it and that's by design.

`rg *` might not look explicit to you, but ripgrep doesn't see that `*`. That `*` is expanded by your shell before invoking ripgrep. So your actual search looks like `rg apple --ignore-file .ignore -in -e 'ohC80hfi'`. You can pretty clearly see this by looking at ripgrep's argv via the `--debug` output:

```
$ echo '/apple' > .ignore
$ echo test > apple
$ rg test
No files were searched, which means ripgrep probably applied a filter you didn't expect.
Running with --debug will show why files are being skipped.
[agallant@agallant-ltl1 rg2250]$ rg test *
1:test
$ rg test * --debug
DEBUG|rg::config|crates/core/config.rs:40: /home/agallant/.ripgreprc: arguments loaded from config file: ["--max-columns-preview", "--colors=match:bg:0xff,0x7f,0x00", "--colors=match:fg:white", "--colors=line:none", "--colors=line:fg:magenta", "--colors=path:fg:green", "--type-add=got:*_test.go"]
DEBUG|rg::args|crates/core/args.rs:543: final argv: ["rg", "--max-columns-preview", "--colors=match:bg:0xff,0x7f,0x00", "--colors=match:fg:white", "--colors=line:none", "--colors=line:fg:magenta", "--colors=path:fg:green", "--type-add=got:*_test.go", "test", "apple", "--debug"]
DEBUG|grep_regex::literal|crates/regex/src/literal.rs:58: literal prefixes detected: Literals { lits: [Complete(test)], limit_size: 250, limit_class: 10 }
DEBUG|globset|crates/globset/src/lib.rs:421: built glob set; 0 literals, 0 basenames, 12 extensions, 0 prefixes, 0 suffixes, 0 required extensions, 0 regexes
1:test
```

**Comment by victoriastuart** [comment]
Hi, thank you for the quick response.  I used the asterisk to tell `rg` to seach everything in the PWD (not realizing it was explicitly including excluded files / directories - oops!).

Follow-on test working as expected - thank you again, for your quick response!

```bash
[victoria@victoria persagen]$ pwd
/mnt/Vancouver/domains/persagen

[victoria@victoria persagen]$ find . -name apple
./apple
./banana/apple

[victoria@victoria persagen]$ cat apple
ohC8Ohfi
akaiv8Ee

[victoria@victoria persagen]$ cat banana/apple 
ohC8Ohfi
akaiv8Ee

[victoria@victoria persagen]$ cat .ignore | grep apple
/apple

[victoria@victoria persagen]$ rg --ignore-file .ignore --color always -in  -e 'ohC8Ohfi'
banana/apple
1:ohC8Ohfi

[victoria@victoria persagen]$ 
```

**Reply by BurntSushi** [reply]
Aye. Note also that `.` or `./` would accomplish "search everything in the CWD."

**Reply by victoriastuart** [reply]
Sorry for returning to this; it relates to my original use (above) of an asterisk in a search.  I thought I had a handle on this - perhaps not?

```bash
[victoria@victoria persagen]$ rg -h | grep recurs
ripgrep (rg) recursively searches the current directory for a regex pattern.

[victoria@victoria persagen]$ rg -l -e vlucas
[victoria@victoria persagen]$ rg . -l -e vlucas
[victoria@victoria persagen]$ rg ./ -l -e vlucas
[victoria@victoria persagen]$ rg * -l -e vlucas
src/register.php
public/register.php
composer.lock
composer.json
[victoria@victoria persagen]$ 
```

**Reply by BurntSushi** [reply]
The `--debug` flag will tell you what rules are being applied.

Remember what I said: `*` is a glob that gets expanded by your shell. ripgrep doesn't see `*`.

**Reply by victoriastuart** [reply]
Thank you for your patience; I know that you are busy.

Output from the `--debug` flag indicated that the `.gitignore` file is automatically read by `rg`, without being explicitly specified.

Subsequent reading reveals that this is the expected behavior:

  https://github.com/BurntSushi/ripgrep/blob/master/GUIDE.md#automatic-filtering=

When I override `.gitignore` using an asterisk in the search - or disable `.gitignore` file by renaming it - I get the expected results.

The link above also describes multiple ignore files, ignore file precedence, etc. which precisely addresses my use cases (use of a `.ignore` file for tailored searches; overriding `.gitignore` entries; ...).

Q.E.D.

**Reply by BurntSushi** [reply]
Yes, they are. By design. :) It might help if you read the guide in the root of this repo.
